### PR TITLE
Feature/branch source

### PR DESCRIPTION
**Pull request by baymac**
The first impl of GitLab Branch Source. The code has lots of bugs. The pr is long so let's do a minimal code review and merge it. After that we will follow jira issue tracking and issue specific prs method.

Test this branch on your local machine with `mvn clean install -DskipTests && mvn hpi:run`.

Test the `Multibranch Pipeline Job` with `GitLab Project` branch source. 

Folder Organisation which is `GitLab Group` job type is broken. Not to be tested in this pr.

### CODE REVIEW

**Review by markjacksonfishing — COMMENTED** [review]
Oh my! Please let us make sure that PR's in the future are in much smaller chunks and accompanied by a corresponding JIRA ticket to show what the acceptance criteria is for the given body of work.
This review will take some time due to its size.

**Comment by baymac** [comment]
@markyjackson-taulia It would have been difficult to create JIRA issues for each task in this code. Maybe 100s of them. Although a JIRA issues for each section of the impl could be done. This is basic impl of branch source so I think it wouldn't be difficult to code review, let's just focus on successful compilation and merging on this one. Thereafter we follow the protocols.

File: pom.xml [hunk 1]
```diff
@@ -108,7 +218,7 @@
         <developer>
             <id>LinuxSuRen</id>
             <name>Zhao Xiaojie</name>
-            <email>parichay.barpanda@gmail.com</email>
+            <email>linuxsuren@gmail.com</email>
```
Comment (markjacksonfishing) [inline pom.xml:226]: I would verify with @LinuxSuRen that is ok to have his email here

**Review by markjacksonfishing — COMMENTED** [review]
Travis is showing failures 🤔

**Comment by baymac** [comment]
There are certain spotbugs error. I need to figure them out.

File: README.md [hunk 2]
```diff
@@ -369,10 +369,10 @@ credentials:
 unclassified:
   gitLabServers:
     servers:
-      - credentialsId: "i<3GitLab"
+      - checkoutCredentialsId: "i<3GitLab"
         manageHooks: true
         name: "gitlab.com"
-        serverUrl: "https://gitlab.com"
+        serverName: "https://gitlab.com"
```
Comment (jetersen) [inline README.md]: name and serverName are super confusing and conflicting. :cry:

**Comment by baymac** [comment]
That was a mistake of IDE. That field didn't mean to change. I'll fix it.
It is still name and serverUrl. In branch source we refer `name` as
'serverName' though.

File: README.md (hunk 2 above)
Comment (baymac) [reply README.md]: I have fixed it.

File: pom.xml (hunk 1 above)
Comment (baymac) [reply pom.xml:226]: This is a typofix, previously I had my email in his section. :sweat_smile:

**Comment by baymac** [comment]
I did a lot of undoing because of the spotbugs and also spaghetti code. In this PR, I have 2 issues to address to - [JENKINS-58261](https://issues.jenkins-ci.org/browse/JENKINS-58261), [JENKINS-58263](https://issues.jenkins-ci.org/browse/JENKINS-58263).

**Comment by baymac** [comment]
I just rebased `develop` onto this branch and forced push.

**Comment by baymac** [comment]
@markyjackson-taulia @justinharringa This PR is ready to be merged. The 58263 and 58261 has been fixed.

**Comment by baymac** [comment]
Spotbugs gives some error:

```
[ERROR] Possible null pointer dereference of GitLabSCMSource.gitlabProject in
io.jenkins.plugins.gitlabbranchsource.GitLabSCMSource.retrieveActions(SCMSourceEvent,
TaskListener) on exception path [io.jenkins.plugins.gitlabbranchsource.GitLabSCMSource,
io.jenkins.plugins.gitlabbranchsource.GitLabSCMSource,
io.jenkins.plugins.gitlabbranchsource.GitLabSCMSource,
io.jenkins.plugins.gitlabbranchsource.GitLabSCMSource] Dereferenced at GitLabSCMSource.java:[line
419]Null value at GitLabSCMSource.java:[line 409]Known null at GitLabSCMSource.java:[line
414]Known null at GitLabSCMSource.java:[line 415] NP_NULL_ON_SOME_PATH_EXCEPTION
[INFO] 
```

But I haven't been able to figure this out yet. I am trying, let me know if anyone can help me fix it.

**Review by markjacksonfishing — APPROVED** [review]
lgtm. Will this be the last big PR?

**Comment by baymac** [comment]
@markyjackson-taulia yes absolutely. :sweat_smile:

File: src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java [hunk 3]
```diff
@@ -249,9 +249,15 @@ public FormValidation doTestConnection(@QueryParameter String serverUrl,
             LOGGER.info("Testing Connection..");
             String privateToken = getToken(serverUrl, credentialsId);
             if (privateToken.equals(UNKNOWN_TOKEN)) {
-                LOGGER.error("Cannot find private token");
-                return FormValidation
-                        .errorWithMarkup(Messages.GitLabServer_credentialsNotResolved(Util.escape(credentialsId)));
+                GitLabApi gitLabApi = new GitLabApi(serverUrl, "");
+                try {
+                    gitLabApi.getProjectApi().getProjects(1, 1);
```
Comment (LinuxSuRen) [inline src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java:260]: Why the params are two `1`?

File: src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java [hunk 4]
```diff
@@ -249,9 +249,15 @@ public FormValidation doTestConnection(@QueryParameter String serverUrl,
             LOGGER.info("Testing Connection..");
```
Comment (LinuxSuRen) [inline src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java]: Suggest adding more context, like `"Testing Connection for GitLab, url:"  + serverUrl`

File: src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java (hunk 3 above)
Comment (jetersen) [reply src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java:260]: This is only bound to go wrong :laughing: Not all instances have kept project with id 1 around.

Comment (baymac) [reply src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java:260]: @casz I told this already. This `getProjects` method tells GitLab Server to return (no. of projects, no. of pages). So whtat getProjects(1,1) tells GitLab is return the latest project in 1 page. If there is no project found it returns `empty list`. If the server is invalid, it raises error. There is no project id used here.

File: src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java (hunk 4 above)
Comment (baymac) [reply src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java]: Fixed in 20e31ae.

File: src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java (hunk 3 above)
Comment (LinuxSuRen) [reply src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java:260]: Add comment on it will be better.

**Comment by baymac** [comment]
I tried adding Pipeline Declarative dependency but it caused a lot of dependency issues. I think it is inessential at the moment, you can test the plugin with `hpi:run` using scripted pipelines (not declarative atm, unless you install it from update center in your jenkins instance).

File: src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java (hunk 3 above)
Comment (baymac) [reply src/main/java/io/jenkins/plugins/gitlabserverconfig/servers/GitLabServer.java:260]: Fixed in a359410.

**Review by markjacksonfishing — APPROVED** [review]

**Review by LinuxSuRen — APPROVED** [review]

**Comment by markjacksonfishing** [comment]
We should be good to merge now @baymac

**Pull request merged**